source/agora/node/Ledger.d

```diff
@@ -542,7 +543,138 @@ unittest
+    Transaction[] makeTransactionForFreezing (
+        KeyPair[] in_key_pair,
+        KeyPair[] out_key_pair,
+        TxType tx_type,
+        Transaction[] prev_txs,
+        const Transaction default_tx)
+    {
+        import std.conv;
+
+        assert(in_key_pair.length == Block.TxsInBlock);
+        assert(out_key_pair.length == Block.TxsInBlock);
+
+        assert(prev_txs.length == 0 || prev_txs.length == Block.TxsInBlock);
+        const TxCount = Block.TxsInBlock;
+
+        Transaction[] transactions;
+
+        // always use the same amount, for simplicity
+        const Amount AmountPerTx = Amount(40_000);
+
+        foreach (idx; 0 .. TxCount)
+        {
+            Input input;
+            if (prev_txs.length == 0)  // refering to genesis tx's outputs
+                input = Input(hashFull(default_tx), idx.to!uint);
+            else  // refering to tx's in the previous block
+                input = Input(hashFull(prev_txs[idx % Block.TxsInBlock]), 0);
+
+            Transaction tx =
+            {
+                tx_type,
+                [input],
+                [Output(AmountPerTx, out_key_pair[idx % Block.TxsInBlock].address)]  // send to the same address
+            };
+
+            auto signature = in_key_pair[idx % Block.TxsInBlock].secret.sign(hashFull(tx)[]);
+            tx.inputs[0].signature = signature;
+            transactions ~= tx;
+
+            // new transactions will refer to the just created transactions
+            // which will be part of the previous block after the block is created
+            if (Block.TxsInBlock == 1 ||  // special case
+                (idx > 0 && ((idx + 1) % Block.TxsInBlock == 0)))
+            {
+                // refer to tx'es which will be in the previous block
+                prev_txs = transactions[$ - Block.TxsInBlock .. $];
+            }
+        }
+        return transactions;
+    }
+
+    KeyPair[] getGenKeyPairs ()
+    {
+        KeyPair[] res;
+        foreach (idx; 0 .. Block.TxsInBlock)
+            res ~= getGenesisKeyPair();
+        return res;
+    }
+
+    KeyPair[] getRandomKeyPairs ()
+    {
+        KeyPair[] res;
+        foreach (idx; 0 .. Block.TxsInBlock)
+            res ~= KeyPair.random;
+        return res;
+    }
+
+    KeyPair[] in_key_pairs;
+    KeyPair[] out_key_pairs;
+    Transaction[] last_txs;
+
+    in_key_pairs = getGenKeyPairs();
+    out_key_pairs = getRandomKeyPairs();
+
+    // generate transactions to form a block
+    void genBlockTransactions (size_t count, TxType tx_type, bool is_valid = true)
+    {
+        foreach (idx; 0 .. count)
+        {
+            auto txes = makeTransactionForFreezing (
+                in_key_pairs,
+                out_key_pairs,
+                tx_type,
+                last_txs,
+                GenesisTransaction);
+
+            txes.each!((tx)
+                {
+                    assert(ledger.acceptTransaction(tx) == is_valid);
+                });
+
+            if (is_valid)
```

Comment:
`is_valid` is always true, because it is never set to false.

Reply:
It's not used here, so we'll have to remove it

Reply:
However, if an error occurs and the block is not created, then the values of these variables must not be changed.

Reply:
Ready for review